### ISSUE TITLE
routing across areas

### ISSUE THREAD

**Issue by andrewharvey** [abrensch/brouter#108]
In particular natural=beach for the trekking profile. Routing around the edge of the way is acceptable.

**Comment by Phyks** [comment]
Would be super useful for routing across squares as well, such as this one https://www.openstreetmap.org/#map=19/48.84147/2.32091 (BRouter takes a huge detour even though riding a bicycle at walking speed is permitted on the square http://brouter.de/brouter-web/#map=18/48.84199/2.32091/OpenStreetMap&lonlats=2.318695,48.841489|2.321465,48.84073).

I know https://makina-corpus.com/blog/metier/2013/moodwalkr-behind-the-scenes and https://makina-corpus.com/blog/metier/2018/calcul-ditineraires-pietons-avec-osrm (in French, "Traversée d'un espace ouvert") which precomputes a visibility graph for such areas and use it in OSRM.

Has anyone other ideas/doc on the way this should be implemented in BRouter?

**Comment by nrenner** [comment]
Some links I collected on the topic:
* [How to Cross Open Spaces? (poster)](http://koenigstuhl.geog.uni-heidelberg.de/publications/2016/Hahmann/poster_threeCols.pdf)
* [Integrating Open Spaces into OSM Routing Graphs for Realistic Crossin… (slides)](https://de.slideshare.net/anitagraser/integrating-open-spaces-into-osm-routing-graphs-for-realistic-crossing-behaviour-in-pedestrian-navigation?ref=https://anitagraser.com/2016/07/12/videos-and-slides-from-fossgis-agit-osgeo-day/)
* [Integrating Open Spaces into OpenStreetMap Routing Graphs for Realistic Crossing Behaviour in Pedestrian Navigation](http://austriaca.at/0xc1aa500e%200x0033ffa5.pdf)
* [User:Maxbe/Routen über Flächen – OpenStreetMap Wiki (de)](https://wiki.openstreetmap.org/wiki/User:Maxbe/Routen_%C3%BCber_Fl%C3%A4chen)

**Comment by poutnikl** [comment]
I guess a small area should be transformed internally to a node, a larger one possibly to a ring of short ways along its circumference.

**Comment by Phyks** [comment]
Thanks @nrenner for the docs! Looking at the links, it seems the approach from https://makina-corpus.com/blog/metier/2013/moodwalkr-behind-the-scenes should work quite well for a first version.

This basically boils down to computing the visibility graph as

![https://makina-corpus.com/blog/metier/2013/copy_of_graphe_visibilite_pres_en.png](Visibility graph)

and comes with the nice feature that the resulting ways can probably be added to the current BRouter routing without any extra work, therefore only requiring pre-processing / `mapcreator` changes

![https://makina-corpus.com/blog/metier/2018/espace-ouvert-moodwalkr/image](Resulting route)

The downside is that the resulting routes are not very realistic, but this is probably not a big deal as
1. It will let BRouter at least cross squares and areas.
2. It is good enough for a first version.
3. Not even Google does it well https://www.google.fr/maps/dir/43.6020158,1.4469147/43.6019744,1.4482346/@43.6018933,1.4475484,19.56z/data=!4m2!4m1!3e2 :)

I'll have a look at the visibility graph computation, see if I can come up with a PoC.

**Comment by Phyks** [comment]
I've just realized the issue has quite deviated a lot from the initial issue.

> In particular natural=beach for the trekking profile. Routing around the edge of the way is acceptable.

Routing around the edge of a surface (closed OSM way) is already implemented, this can be seen for squares in http://brouter.de/brouter-web/#map=19/48.81870/2.31914/standard&lonlats=2.319464,48.818572|2.319746,48.81886|2.319909,48.819143. I think the issue for not being able to route around a polygon with `natural=beach` (typically https://www.openstreetmap.org/way/169161621#map=17/46.69360/-1.94501) is that they are not included in the segments file (as they don't appear in https://github.com/abrensch/brouter/blob/master/misc/profiles2/lookups.dat).


So there are a few (quite related) issues:

- [ ] One should probably add `natural=beach` (and other such relevant tags) to allow routing around them (as long as this is a single polygon, made from a closed OSM way).
- [ ] There is the question of routing **across** areas (where an area is any polygon / closed OSM way), which can be achieved quite easily by computing the visibility graph of the polygon and adding new edges to the segment files. This would allow a route across the square in http://brouter.de/brouter-web/#map=19/48.81857/2.31943/standard&lonlats=2.319593,48.818601|2.319373,48.819214.
- [ ] Finally, one should also add support for the multipolygons (defined as OSM relations with outer and inner boundaries), which would allow routing across complex squares such as http://brouter.de/brouter-web/#map=18/48.84199/2.32091/standard&lonlats=2.318695,48.841489|2.321465,48.84073.

**Comment by zossebart** [comment]
>     * There is the question of routing **across** areas (where an area is any polygon / closed OSM way), which can be achieved quite easily by computing the visibility graph

slightly offtopic: if this happens to get implemented, maybe the visibility graph computation could be implemented in a reusable way.
With this, maybe something similar to this could be implemented later: [http://k1z.blog.uni-heidelberg.de/2018/11/07/generating-customized-pleasant-pedestrian-routes-based-on-openstreetmap-data/](url)
Especially the "greenness" sounds interesting and is also implemented using a visibility graph.

Maybe I should start a separate issue/thread regarding this, though...

**Comment by Phyks** [comment]
I could chat about the visibility graph approach with people working with it at the SOTM at Heidelberg (https://media.ccc.de/v/sotm2019-1359-pedestrian-routing-in-complex-areas-the-case-of-paris-railway-stations) and it seems quite tricky to have it working well actually. The main issue being that computing the visibility graph can be very power consuming, especially when you get a lot of nodes in your surface. With a naive algorithm (complexity `O(n³)`), it takes them about 20 minutes to ingest a small area around Gare de l'Est in Paris https://www.openstreetmap.org/#map=17/48.87781/2.35990 (with many surfaces with a few hundreds or thousands points).

There seems to be more refined algorithms with lower complexity (http://cs.smith.edu/~istreinu/Teaching/Courses/274/Spring98/Projects/Philip/fp/visibility.htm) but the complexity and implications of the area routing should probably be carefully investigated before starting to work on this.

**Comment by codemewell** [comment]
Maybe this will be helpful. Here's working example: https://turfjs.org/docs/#shortestPath

It uses A* algoritm. Has obstables input. Matrix resolution.

No idea which approach from mentioned above is this based on. But if it works in popular library maybe is worth considering?

**Comment by nrenner** [comment]
There also was this talk at SotM: 
[Client-side route planning: preprocessing the OpenStreetMap road network for Routable Tiles](https://2019.stateofthemap.org/sessions/GRJC7W/).

They implement a visibility graph and then reduce the number of paths to those actually taken by the router (and use it as an example where it would make sense to share the results of such expensive preprocessings): [video (5:17-)](https://youtu.be/_hH8qLEwL6M?t=317), [slides (10,11)](https://hdelva.be/slides/sotm2019/#/10).

**Comment by polyscias** [comment]
See https://github.com/Project-OSRM/osrm-backend/pull/7161, OSRM added support for area routing. 

Seems quite doable as pre-processing step.

**Comment by afischerdev** [comment]
@polyscias 
Great. Thanks for the hint. 
I had in the past some test with areas. 
But lua scripts could be a new idea. And we had a little contact with this script format when we started with pseudo tags.
My starter idea was to 
- collect the polygon for the area
- collect relevant nodes for entry/exit
- build the conections between this nodes
- collect all inside polygons or ways for the area
- build the transfer nodes with this

The question always was `what to do with the inside ways`?
Add the direct way from entry to exit to the present way?
Drop it when already a candidate is present?
Connect direct to an existing ways?

E.g. overpass around city of Darmstadt
```
way["highway"="pedestrian"]["area"="yes"]({{bbox}});
way["highway"="pedestrian"]({{bbox}});
```

**Comment by polyscias** [comment]
On how to implement this, if I read the linked OSRM issue I see they use a visibility graph

![Image](https://github.com/user-attachments/assets/975cdcf9-5adf-4685-a562-465e559662b8)

I see there is support to calculate that in standard open source libraries:

https://grass.osgeo.org/grass-stable/manuals/v.net.visibility.html

So I would probably use that.

The next step is to run then a routing algorithm on this visibility graph from each entry point for every combination of points and leave only the segments visited in that process, also something that can be done using grass and some glue code. Finally adds those remaining edges as ways to the the input for brouter .rd5 creation.

> The question always was what to do with the inside ways?

You can safely drop them but if that is harder to implement (paths continuing over the edges of the area) there is no problem leaving them, the router will likely not choose them as the routes found via the visibility graph are the shortest.

The question is more on what to do with tags on the crossing ways if they are not uniform.
